File: hashing/subarray_sum_equals_k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @returns {number}
 */
var subarraySum = function(nums, k) {
    let counts = new Map();
    counts.set(0, 1);

    let ans = 0;
    let curr = 0;

    for (const num of nums)
    {
        curr += num;
        ans += counts.get(curr - k) || 0;
        counts.set(curr, (counts.get(curr) || 0) + 1);
    }

    return ans;
}

File: hashing/subarray_sum_equals_k.py
from collections import defaultdict

def subarraySum(nums: List[int], k: int) -> int:
    counts = defaultdict(int)
    counts[0] = 1

    ans = curr = 0
    for num in nums:
        curr += num
        ans += counts[num - k]
        counts[curr] += 1
    
    return ans
